crates/fuel-core/src/service/sub_services/algorithm_updater.rs

```diff
@@ -0,0 +1,209 @@
+    tables::{
+        FuelBlocks,
+        Transactions,
+    },
+    transactional::HistoricalView,
+    StorageAsRef,
+};
+use fuel_core_types::{
+    fuel_tx::field::MintAmount,
+    fuel_types::BlockHeight,
+    services::block_importer::SharedImportResult,
+};
+use std::cmp::Ordering;
+
+type Updater = FuelGasPriceUpdater<
+    FuelL2BlockSource<ConsensusParametersProvider>,
+    StructuredStorage<Database<GasPriceDatabase, RegularStage<GasPriceDatabase>>>,
+>;
+
+pub fn get_synced_gas_price_updater(
+    config: &Config,
+    genesis_block_height: BlockHeight,
+    settings: ConsensusParametersProvider,
+    block_stream: BoxStream<SharedImportResult>,
+    mut gas_price_db: Database<GasPriceDatabase, RegularStage<GasPriceDatabase>>,
+    on_chain_db: Database<OnChain, RegularStage<OnChain>>,
+) -> anyhow::Result<Updater> {
+    let metadata_height: u32 = gas_price_db
+        .latest_height()?
+        .unwrap_or(genesis_block_height)
+        .into();
+    let latest_block_height: u32 = on_chain_db
+        .latest_height()?
+        .unwrap_or(genesis_block_height)
+        .into();
+
+    let genesis_metadata = UpdaterMetadata::V0(V0Metadata {
+        new_exec_price: config.starting_gas_price,
+        min_exec_gas_price: config.min_gas_price,
+        exec_gas_price_change_percent: config.gas_price_change_percent,
+        l2_block_height: genesis_block_height.into(),
+        l2_block_fullness_threshold_percent: config.gas_price_threshold_percent,
+    });
+    if BlockHeight::from(latest_block_height) == genesis_block_height {
+        revert_gas_price_db_to_height(&mut gas_price_db, latest_block_height.into())?;
+        let metadata_storage = StructuredStorage::new(gas_price_db);
+        let l2_block_source =
+            FuelL2BlockSource::new(genesis_block_height, settings, block_stream);
+        let updater = FuelGasPriceUpdater::new(
+            genesis_metadata.into(),
+            l2_block_source,
+            metadata_storage,
+        );
+        Ok(updater)
+    } else {
+        let metadata_storage = match metadata_height.cmp(&latest_block_height) {
+            Ordering::Equal => StructuredStorage::new(gas_price_db),
+            Ordering::Less => {
+                let mut metadata_storage = StructuredStorage::new(gas_price_db);
+                sync_metadata_storage_with_on_chain_storage(
+                    &settings,
+                    &mut metadata_storage,
+                    on_chain_db,
+                    metadata_height,
+                    latest_block_height,
+                    genesis_metadata,
+                    genesis_block_height.into(),
+                )?;
+                metadata_storage
+            }
+            Ordering::Greater => {
+                revert_gas_price_db_to_height(
+                    &mut gas_price_db,
+                    latest_block_height.into(),
+                )?;
+                StructuredStorage::new(gas_price_db)
+            }
```

Comment:
I don't like how complex and special-casey this code is. A lot of the complexity is coming from the fact that we don't store the metadata for the genesis block, so we have to manually pass the `genesis_metadata` around for certain cases.

I also think calling `revert_gas_price_db_to_height` in two different locations is a smell, so I might sleep on this code and see if a clear approach arises. Kinda been staring at it too long to think straight.